Extended 'mix' type functionality

muppet3000 opened the pull request:
- Added API calls for Mix Totals & Current system status
- Fixed the `mix_info` function as it was missing the `mix_id` parameter
- Full example added for a 'mix' type system
- Updated the README accordingly

@indykoning - The intention once this is merged is to ask you to do a formal release so that I can then add the Mix functionality to the HomeAssistant integration.

muppet3000 commented:
@indykoning any chance we could get this merged in & released so I can integrate my Growatt system into Home Assistant?

muppet3000 commented:
This fixes issue: https://github.com/indykoning/PyPi_GrowattServer/issues/5

indykoning commented:
Absolutely, once the review has been solved i'll be happy to merge it!

muppet3000 commented:
I think you need to approve it? 
If there's something you need me to do to get it ready please let me know.

indykoning reviewed (changes requested):
I hadn't finished the review so it just left it as a comment on the code, you should be able to see it now 🙂

muppet3000 commented on growattServer/__init__.py:
```diff
@@ -144,17 +144,44 @@ def tlx_detail(self, tlx_id):
         return data
 
-    def mix_info(self, mix_id):
+    def mix_info(self, plant_id, mix_id):
```
I'll test it today and get back to you. I'll post the results of each of the options.

I was worried about introducing an interface breaking change as well. If it doesn't work in either scenario then there's an argument to say that it isn't interface breaking it's actually a bugfix :P

muppet3000 replied:
See latest changes

muppet3000 commented:
@indykoning I've investigated and you're correct it does work when you just pass the serial number in. However, the other functions I've added need the plant ID to be provided.

The latest change I have pushed modifies the ordering of the parameters and makes the `plant_id` for the `mix_info` function optional. Therefore, if it becomes a required parameter on the API in the future it should be easy for people to adopt rather than requiring another change to this library.

I have tested that it is backwards compatible and updated the documentation accordingly.

muppet3000 commented:
@indykoning are you happy to merge this in now?

muppet3000 commented:
@indykoning I've pushed my final changes now after doing more investigations on the correct values from my end. 
I've added the best documentation that I can manage for the api calls that I've added by trying to work out on the app where the values are presented. If I find more in the future I'll update them.

Please can you review, merge and release this soon so that I can get on with the home-assistant integration?

muppet3000 commented:
@indykoning Please could you review this so that I can move on to the HomeAssistant integration?

muppet3000 commented on growattServer/__init__.py:
```diff
@@ -144,17 +144,219 @@ def tlx_detail(self, tlx_id):
+        #Therefore, here we take a moment to calculate the same values provided elsewhere but based on the graph data instead
+        #The particular stats that we question are 'load consumption' (elocalLoad)  and 'import from grid' (etouser) which seem to be calculated from one-another
+        #It would appear that 'etouser' is calculated on the backend incorrectly for systems that use AC battery charged (e.g. during cheap nighttime rates)
+        if timespan == Timespan.hour or timespan == Timespan.day:
```
I'm happy to remove it and put it into the HomeAssistant library and my example file only. 

HOWEVER, note that I do not modify any of the values that are returned, I just add new ones to the data structure that is returned from the API. I thought it would be helpful for people to have this done for them, they can then choose whether or not they want to use it.

What would you like me to do? I'm happy either way, I'd just like to get it merged in :)

muppet3000 replied:
@indykoning any thoughts on how you'd like to proceed on this one?

indykoning replied:
Id like to keep any custom code at a minimum unless it's for backwards compatible reasons

muppet3000 replied:
No problem at all. I completely understand. 

I'll remove these additions to the function and make it just return the raw values with no additions. I'll move this logic into my example file so that others can see how to calculate these values if they wish.

muppet3000 commented:
@indykoning latest changes from our discussion now pushed, no manual additions to the raw data from the API.
I have moved the calculation to the example so that others can use the same method if they wish to.

muppet3000 commented:
@indykoning are you happy with all the functionality now?

indykoning commented:
At a first glance it looks good, i'll have a last check and will try to release it when i have my computer which may be tomorrow

muppet3000 commented:
Thanks, sounds great. If you're able to release it tomorrow that would be amazing I could then start on the additions to HA integration over the weekend.

muppet3000 commented:
@indykoning please don't merge it until you're going to release tomorrow, there's a very slim chance I've found the final statistic I've been looking for on the API. I need to verify it when I'm at my desk in about 2 hours time. 
If it's the stat I think it is then I'll add it in asap and give you an update.

muppet3000 commented:
@indykoning I've now added that final api call and fully documented it. It's now down to you for final review, merge & release.

Thanks for waiting.

indykoning commented:
It's looking good :tada: 
I'll merge it and i think i'll release it as a major version just in case.
So it'll be 1.0.0 :rocket:

indykoning commented:
It is available at:
https://pypi.org/project/growattServer/1.0.0/

muppet3000 commented:
Awesome, thanks @indykoning it's been fun working with you in this one. I'll probably start on the HomeAssistant integration tomorrow. Fun times!
